### PR TITLE
#773 boardpicker bug

### PR DESCRIPTION
**Pull request by derimmer**

### CODE REVIEW

File: src/VASL/build/module/map/boardPicker/ASLBoard.java [hunk 1]
```diff
@@ -129,21 +129,24 @@ public void initializeFromArchive(File archiveFile) {
...
+            imageFile = VASLBoardArchive.getBoardImageFileName();
+            version = VASLBoardArchive.getVersion();
+
+            ((Translatable) getGrid()).setAttribute(HexGrid.X0, (int) VASLBoardArchive.getA1CenterX());
+            ((Translatable) getGrid()).setAttribute(HexGrid.Y0, (int) VASLBoardArchive.getA1CenterY());
+            ((Translatable) getGrid()).setAttribute(HexGrid.DX, VASLBoardArchive.getHexWidth());
+            ((Translatable) getGrid()).setAttribute(HexGrid.DY, VASLBoardArchive.getHexHeight());
+            ((Translatable) getGrid()).setAttribute(HexGrid.SNAP_SCALE, VASLBoardArchive.getSnapScale());
+            ((Translatable) getGrid()).setAttribute(BoardArchive.ALT_HEX_GRID_KEY, Boolean.toString(VASLBoardArchive.isAltHexGrain()));
+        } catch (Exception e) {
```
Comment (uckelman) [inline src/VASL/build/module/map/boardPicker/ASLBoard.java:146]: You should almost never catch `Exception`. The right way to fix the NPE here is not to cause it. You need to check if `VASLBoardARchive` is null and then proceed accordingly.

**Comment by derimmer** [comment]
I tried that at first. However, I was able to generate errors where the VASLBoardArchive was created but with missing values. So, I used try catch to trap those errors. It also catches VASLBoardArchive = Null. 

If the VASLBoardArchive contains missing or erroneous data, there is no way to fix that in mid-process.

**Comment by uckelman** [comment]
I can't stress enough how poor this approach is. When you catch `Exception`, you're catching everything that could be coming out of that block of code, which includes bugs you don't know about right now---and then by having nothing but a `return` inside the catch block, you're throwing away all the evidence of what happened. By doing this, you make it difficult to find bugs in any of the code inside the try. (There are a few rare cases where catching `Exception` is appropriate, but they're mainly to do with reflection or badly-behaving third-party libraries, and the try blocks are typically restricted to single lines in those cases. This is not one of those cases.)

Many code analysis packages (e.g., checkstyle, PMD, spotbugs) have rules against catching `Exception` and will complain. If I saw this in a code review at work, I would make whoever wrote this code fix it. No PR with this in it would ever be merged.

Sometimes the correct thing to do is to narrow the exception type you're catching, but that's never the right thing to do when the exception you want to catch is a `NullPointerException`. You should _never_ use NPEs as flow control. They are intended in all cases to indicate a logic error in the program. (Many code analysis packages have rules against this as well.)

The right thing to do in this case is the following:

```
if (VASLBoardArchive == null) {
  return;
}

setCommonName(VASLBoardArchive.getBoardName());
imageFile = VASLBoardArchive.getBoardImageFileName();
version = VASLBoardArchive.getVersion();

((Translatable) getGrid()).setAttribute(HexGrid.X0, (int) VASLBoardArchive.getA1CenterX());
((Translatable) getGrid()).setAttribute(HexGrid.Y0, (int) VASLBoardArchive.getA1CenterY());
((Translatable) getGrid()).setAttribute(HexGrid.DX, VASLBoardArchive.getHexWidth());
((Translatable) getGrid()).setAttribute(HexGrid.DY, VASLBoardArchive.getHexHeight());
((Translatable) getGrid()).setAttribute(HexGrid.SNAP_SCALE, VASLBoardArchive.getSnapScale());
((Translatable) getGrid()).setAttribute(BoardArchive.ALT_HEX_GRID_KEY, Boolean.toString(VASLBoardArchive.isAltHexGrain()));
```

This way, you need no try-catch and if there's a bug in the code you have a chance of seeing the exception.

**Comment by derimmer** [comment]
Thanks for this. I am a non-professional, self-taught coder and so I lack a lot (if not all) of this good practice knowledge. However, I am capable of learning so I am always glad to get advice. I will revise the code accordingly. However, it won't catch the particular problem that revealed this bug, which was that while VASLBoardArchive was set the version number property was not and thus version=VASLBoardArchive.getVersion() generated an NPE. 

Seems to me there are two ways to address this. One would be to revise VASLBoardArchive so it returns null if it is missing values (or to use some kind of placeholder values). The Second would be to wrap each of the code statements that follow the if (VASLBoardArchive == null) test in their own is null test. For example, VASLBoardArchive.getBoardName() == null ? setCommonName("") : setCommonName(VASLBoardArchive.getBoardName());

Which would you suggest? Or is there a better way?

**Comment by uckelman** [comment]
I'll have a look this evening.

Note that I'm not aiming to give you a hard time here, just trying to make you aware of some best practices.

**Comment by uckelman** [comment]
It occurs to me that one or more of the checkers I mentioned would help you guys considerably. I'll have to show you how to set them up.

**Comment by derimmer** [comment]
> I'll have a look this evening.
> 
> Note that I'm not aiming to give you a hard time here, just trying to make you aware of some best practices.

I do not have a thin skin when it comes to coding. I know I know nothing. All advice greatly appreciated. LOL!

**Comment by derimmer** [comment]
@uckelman when I remove the first try catch, it produces an error message that an IOException is not being handled. So, I have put the try catch back but have followed it with if (VASLBoardArchive == null) {return;}. 

`try{
      VASLBoardArchive = new BoardArchive(archiveFile.getName(), archiveFile.getParent(), ASLMap.getSharedBoardMetadata());
 } catch (IOException e) {
            ErrorDialog.dataWarning(new BadDataReport("Unable to open board file", archiveFile.getName(), e));
 }
 if (VASLBoardArchive==null){return;}`

Does that work?

That leaves the second try catch. I should replace it with tests for each of the values being set?

**Comment by uckelman** [comment]
> @uckelman when I remove the first try catch, it produces an error message that an IOException is not being handled. So, I have put the try catch back but have followed it with if (VASLBoardArchive == null) {return;}.
> 
> `try{ VASLBoardArchive = new BoardArchive(archiveFile.getName(), archiveFile.getParent(), ASLMap.getSharedBoardMetadata()); } catch (IOException e) { ErrorDialog.dataWarning(new adDataReport("Unable to open board file", archiveFile.getName(), e)); } if (VASLBoardArchive==null){return;}`
> 
> Does that work?

Yes, but you could simply return from the catch block, because you already know that the `VASLBoardArchive` is null inside the catch block, and it's definitely not null if you're after the catch block.

Do what @zgrose suggested on #811.

**Comment by derimmer** [comment]
OK. I have done. Hard to know if it works since I cannot ever replicate the error. Somehow during the board selection process, selected boards are not opening. I have no idea why. We can but try. I am going to close the numerous bug reports on VASSAL about this error and #811.

**Pull request merged**